### DISCUSSION TITLE
Inquiry Regarding Long Computation Time for Jacobian and Residual Calculation

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#27307] [Q&A Modules: Solid mechanics]
Dear Moose Experts,

I am currently attempting to replicate a crystal plasticity-phase field coupled model as described in the paper "[On the plastic driving force of grain boundary migration: A fully coupled phase field and crystal plasticity model](https://www.sciencedirect.com/science/article/abs/pii/S092702561630605X)". The authors have claimed that this model has already been implemented in MOOSE/MARMOT. However, in my own implementation of the crystal plasticity-phase field coupled model, I have encountered prolonged computation time for Jacobian and residual calculation when the simulation reaches "Time Step 75, time = 3.27713, dt = 0.0855946", compared to the time steps before.
![image](https://github.com/idaholab/moose/assets/56183775/64e54997-76ca-4f8e-946a-29c1d01fdc32)

Specifically, according to the description of the coupling method outlined by the authors (see Sec 2.4 coupling method), at the grain boundaries or interfaces, it is necessary to compute individual crystal plasticity models for each activated grain or order parameter (GrainTracker), followed by the construction of kernels (StressDivergenceTensors) using the weighted average Cauchy stress and Jacobian matric.

To achieve this coupling model, based on the crystal plasticity models`ComputeElasticityTensorCP`, `ComputeMultipleCrystalPlasticityStress`, and `CrystalPlasticityKalidindiUpdate` provided in the example file [exception.i](https://github.com/idaholab/moose/blob/4781fcaf79a66ea8012b06da93aa744d283d2e87/modules/solid_mechanics/test/tests/crystal_plasticity/stress_update_material_based/exception.i#L4), I have extended the paradigm to consider multiple crystal plasticity models for each activated grain at every quadrature point. The specific construction approach is illustrated in the attached diagram,
![image](https://github.com/idaholab/moose/assets/56183775/eac8e3ce-3d62-498c-b97a-04d53edd1f35)

The main idea is to extend the state parameters in the crystal plasticity constitutive model to vector types, where the size of the vector-type material parameters corresponds to the number of order parameters. The material parameters associated with each grain include elastic tensor, rotation matrix, plastic deformation gradient, second Piola-Kirchhoff stress, Cauchy strain, and Jacobian matrix. The existing material parameters in MOOSE's crystal plasticity constitutive model are replaced with the aforementioned material parameters indexed by order parameters. At each quadrature point, after computing the crystal plasticity models corresponding to each activated order parameter, weighted Cauchy stresses _stress[_qp] and Jacobian matrices _Jacobian_mult[_qp] are used to construct the kernel based on the weighting method proposed in the paper.

In summary, the above is my approach to constructing the coupled model based on the coupling method described in the paper. The current issue I am facing is the prolonged computation time for Jacobian and residual calculation at approximately 3.3 seconds, where the run time is 1552.32 seconds, whereas before 3.3 seconds, the run time was less than 5.0 seconds. I am seeking assistance in understanding the reasons behind the extended computation time at "Time Step 75".

I would greatly appreciate any insights or suggestions from the community to help resolve this issue.

Thank you very much for your time and consideration.

Best regards,
Wei

**Comment by GiudGiud** [comment]
Hello

so you are suspecting the material property computation is slow?
Would you mind generating a profile with a profiler? There is a page explaining how to do that on the moose website. 

Guillaume

**Reply by PengWei97** [reply]
Hi @GiudGiud,

Thanks for your quick reply and advice.

> so you are suspecting the material property computation is slow?

It's not actually slow when calculating material properties, because I try to insert the statement in the function `ComputeMultiCPStressFullyCoupled::computeQpStress()`, in fact the statement can run through quickly, but stays at `Computing Residual..... .....................`, so now I am mainly not sure where the problem lies, and it takes so long to display the calculation residual or Jacobian matrix.

> Would you mind generating a profile with a profiler? There is a page explaining how to do that on the moose website.

I would be happy to create a profile to help me find where the problem lies. Can you provide me with a link to this web page? At the same time, I am also happy to share my current semi-finished code, which is actually based on the existing classes related to crystal plasticity.

Wei

**Reply by GiudGiud** [reply]
hello

the page for profiling is here
https://mooseframework.inl.gov/application_development/profiling.html

> in fact the statement can run through quickly, but stays at Computing Residual..... .....................

interesting. I think profiling will tell us more about what is happening

**Reply by PengWei97** [reply]
Hi @GiudGiud,

Following your suggestion, I attempted to build and utilize gperftools to identify functions with long running times. However, despite spending considerable time configuring the environment based on the webpage you provided, I encountered an error even after running the command `MOOSE_PROFILE_BASE=run1_ mpiexec -n 36 ~/projects/panda/panda-oprof -i s2_bicrystal_cp_elastic3.i > 01.log`. Subsequently, when I used `pprof run1.prof`, I still encountered the following error:

```bash
$ pprof run1.prof
Using remote profile at run1.prof.
Use of uninitialized value $line in substitution (s///) at /home/pw-moose/miniforge/envs/moose/bin/pprof line 3326.
http://run1.prof/pprof/symbol doesn't exist
```

Therefore, regarding the issue with setting up the environment, could you please provide some advice? Alternatively, if possible, could you run the example [s2_bicrystal_cp_elastic3.i](https://github.com/PengWei97/panda/commit/cc3afebfe5197ed43345c3852391075ca8d2007e#diff-439822fd9966857d9363f08c92736c907c3037ebc1211854a7def320237923bf) in the panda::prm-fully environment to identify where the long running time occurs? Of course, I could also email you the files run1_0.prof through run1_35.prof.

Best regards,
Wei

**Reply by PengWei97** [reply]
Furthermore, I attempted to set the parameter 'ao = 0' in `CPKalidindiFullyCoupledUpdate`, which effectively disables plasticity behavior in the constitutive model, leaving only the elastic Hookean behavior. Strangely, the simulation of crystal plasticity-phase field coupling proceeded as expected, with grain boundaries migrating under elastic driving forces, as shown below:
![image](https://github.com/idaholab/moose/assets/56183775/b23c89a7-c033-40b0-8102-03ce6edb5dec)

However, when setting 'ao = 0.001' to activate plasticity in the constitutive model, the simulation still encountered a significant computational process during the time step calculation, indicated by the display of 'Residual' in the terminal as follows:
```
Time Step 8, time = 0.108816, dt = 0.020736

Grain Tracker Status:
Grains active index 0: 1 -> 1
Grains active index 1: 1 -> 1

Finished inside of GrainTracker

    Computing Residual.......................................................................................................................................................................................................................................................................................................................................................
```

Therefore, I suspect that the need for **long-duration residual calculations arises from the plastic constitutive model in the mechanical model**. However, I currently lack specific insights into why this situation occurs. I earnestly request advice from MOOSE experts. 

Alternatively, I believe it may be necessary to re-examine the coupling strategy and modeling methods outlined in the Zhao's paper. Of course, any guidance or suggestions regarding the modeling of crystal plasticity-phase field coupling would be greatly appreciated.

Best regards,
Wei

**Reply by GiudGiud** [reply]
wrt to your first message, I wont have time for another two weeks at least to profile that for you. 

> panda::prm-fully environment 

you mean the prm-fully branch of panda?

@laagesen thoughts on the last message?

**Reply by PengWei97** [reply]
> wrt to your first message, I wont have time for another two weeks at least to profile that for you.

Okay, thank you very much for your attention to this. I will also try to solve this problem in the two weeks of class. If it is not solved yet, I will also let you know the news.

> you mean the prm-fully branch of panda?

Yes, I have placed the complete the input file and the new classes I created into my own app, panda, in the branch prm3-fully.

**Reply by PengWei97** [reply]
Hi @GiudGiud,

It is very regrettable to say that we have not yet obtained the performance analysis report of the program running. In the past two weeks, we have successfully configured the pprof environment  and now have encountered new problems #27467.

Given these circumstances, I am reaching out to see if you might have the time and willingness to assist me with this task. Your expertise and insights would be invaluable in helping me navigate these challenges, and any assistance you could provide in profiling the program for me would be greatly appreciated.

Thank you for considering my request. I am looking forward to your favorable response and am hopeful for your support.

Warm regards,

Wei

**Reply by GiudGiud** [reply]
why are you using 32 processes btw? I just run your input and it was fine with 2

With 2, the top is
```
Welcome to pprof!  For help, type 'help'.
(pprof) Total: 2773 samples
     187   6.7%   6.7%      192   6.9% RankFourTensorTempl::operator*@101fb94
     149   5.4%  12.1%      149   5.4% 0x0000000184d2ab70
     141   5.1%  17.2%      145   5.2% RestartableDataReader::deserializeValue
     121   4.4%  21.6%      121   4.4% 0x0000000184d22de0
      73   2.6%  24.2%       73   2.6% 0x0000000184d92f5c
      61   2.2%  26.4%       89   3.2% RankTwoTensorTempl::times@a72c
      60   2.2%  28.6%       60   2.2% 0x0000000184d260c4
      59   2.1%  30.7%       74   2.7% CPKalidindiFullyCoupledUpdate::calculateStateVariableEvolutionRateComponent
      54   1.9%  32.6%       54   1.9% RankFourTensorTempl::operator+=
      44   1.6%  34.2%       44   1.6% libMesh::Parameters::Parameter::print@a6a140
(pprof) 
```
some symbols are not resolved. This is likely due to my computer being hosed with 'security' stuff. It should work fine for yours

**Reply by PengWei97** [reply]
Thank you very much for your efforts. 

Ah ha, that's weird. Even if I use 2 cores for parallel computing, the persistence of "Computing Residual..." still occurs at `Time Step 8, time = 0.108816, dt = 0.020736`, as follows:
```
Time Step 8, time = 0.108816, dt = 0.020736

Grain Tracker Status:
Grains active index 0: 1 -> 1
Grains active index 1: 1 -> 1

Finished inside of GrainTracker


     Computing Residual.............
```
I use the configuration of `788258e49531eb24bfd347a600d69a16f966c495` historical node. So what I want to ask is, did you run to this simulation time, that is, time = 0.108816, and did it pass?

**Reply by GiudGiud** [reply]
ok I only profiled the first two time steps.
I am seeing it stuck on time step 2

**Reply by PengWei97** [reply]
Yes, in fact, this simulation runs well in the first seven steps, but during the eighth step of simulation, "Computing Residual............." will continue to be displayed, and My goal is to solve this problem.

**Reply by GiudGiud** [reply]
Ok I used a debugger and interrupted during the very long time step 8 and the slow thing is computing the material
```
  * frame #0: 0x000000010077dbd4 libpanda-devel.0.dylib`libMesh::TypeTensor<double>::operator()(this=<unavailable>, i=1, j=1) const at type_tensor.h:718:3 [opt]
    frame #1: 0x0000000100786488 libpanda-devel.0.dylib`libMesh::TypeTensor<libMesh::CompareTypes<double, double>::supertype> libMesh::TypeTensor<double>::operator*<double>(this=0x00000001450b6de0, p=0x000000016fdf9e50) const at type_tensor.h:1239:27 [opt]
    frame #2: 0x000000010079016c libpanda-devel.0.dylib`ComputeMultiCPStressFullyCoupled::calculateResidual() + 388
    frame #3: 0x000000010078f124 libpanda-devel.0.dylib`ComputeMultiCPStressFullyCoupled::solveStress() + 76
    frame #4: 0x000000010078ee68 libpanda-devel.0.dylib`ComputeMultiCPStressFullyCoupled::solveStateVariables() + 108
    frame #5: 0x000000010078e8c0 libpanda-devel.0.dylib`ComputeMultiCPStressFullyCoupled::solveQp() + 128
    frame #6: 0x000000010078e4e8 libpanda-devel.0.dylib`ComputeMultiCPStressFullyCoupled::updateStress(RankTwoTensorTempl<double>&, RankFourTensorTempl<double>&) + 712
    frame #7: 0x00000001007962c8 libpanda-devel.0.dylib`ComputeMultiCPStressFullyCoupled::computeQpStress() + 588
    frame #8: 0x00000001066e4dc8 libsolid_mechanics-devel.0.dylib`ComputeGeneralStressBase::computeQpProperties(this=0x00000001450b5a18) at ComputeGeneralStressBase.C:46:3 [opt]
    frame #9: 0x00000001037af998 libmoose-devel.0.dylib`Material::computeProperties(this=0x00000001450b5a18) at Material.C:139:7 [opt]
    frame #10: 0x0000000103e7192c libmoose-devel.0.dylib`FEProblemBase::reinitMaterials(unsigned short, unsigned int, bool) at MaterialData.h:313:10 [opt]
    frame #11: 0x0000000103e71910 libmoose-devel.0.dylib`FEProblemBase::reinitMaterials(this=0x0000000146010818, blk_id=<unavailable>, tid=<unavailable>, swap_stateful=<unavailable>) at FEProblemBase.C:3528:21 [opt]
    frame #12: 0x0000000104477c58 libmoose-devel.0.dylib`NonlinearThread::onElement(this=0x000000016fdfaf60, elem=<unavailable>) at Nonl
```

so the way forward is for you to speed that material up.
look for unnecessary or slow operations, especially memory allocations can be very expensive

**Reply by PengWei97** [reply]
Hi @GiudGiud,

Yes, I suspect that the problem may stem from improper configurations in the ComputeMultiCPStressFullyCoupled Class. As mentioned previously, I have expanded the state parameters in the crystal plasticity model at each integration point based on the number of active order parameter (in phase field). Subsequently, I calculate the stress and Jacobian_mult through weighting, which are then utilized in the kernel::StressDivergenceTensors for residual and Jacobian computations.

It's worth noting that the simulation runs smoothly when considering only the elastic part. However, the issue arises when incorporating the plasticity component, specifically stalling at step 8 with the continuous "Computing Residual" message.

So I have been looking for the reason why `Computing Residual.............` continues to display. This code shows that it is executed in `moose/framework/src/problems/FEProblemBase.C`. I would greatly appreciate any insights or any guidance from you, @sapitts, and @jiangwen84. I am confident that the collective expertise of yours could provide valuable assistance in resolving this issue.

If this problem appears trivial or if I have overlooked any crucial details, I apologize for any inconvenience caused by my lack of expertise in this area.

Thank you for your attention to this matter.

Best regards,
Wei

**Reply by GiudGiud** [reply]
> So I have been looking for the reason why Computing Residual............. continues to display. This code shows that it is executed in moose/framework/src/problems/FEProblemBase.C

which code? which lines?
When I was using the debugger and interrupting it regularly, lines in `ComputeMultiCPStressFullyCoupled` were showing up instead

**Reply by PengWei97** [reply]
Yes, you are right. I was misled for a long time by `TIME_SCRTION("computeResidualTags", 5, "Computing Residual");` in FEProblemBase. In fact, TIME_SECTION is a performance prompt.

After that, I tried to add on some output statements to `ComputeMultiCPStressFullyCoupled::computeQpStress()`, and the result output to the terminal is as shown in the below figure:

![image](https://github.com/idaholab/moose/assets/56183775/010feef4-8226-44bc-a92f-af5a59feec01)

The results prove that your are right. When the number of activation order parameters is greater than 2, that is, at the grain boudary, each element requires a long time to calculate. This is the range that I can currently narrow down.

**Comment by PengWei97** [comment]
Hi @GiudGiud, @jiangwen84, and @sapitts,

Using the strategy proposed in #27468, I got a profiling file for this problem and the results are as follows:
![run_1](https://github.com/idaholab/moose/assets/56183775/624b5e68-c750-4b27-ac7f-5efdacf46a4e)

The results show that a long time is spent `ComputeMultiCPStressFullyCoupled::calculateJacobian()`. After that, I will carefully chech which statement takes a long time to calculate and what is the  reasion. Additionally, I am guessing that the fetch may be caused by inappropriate tensor values.

Finally, I would greatly appreciate any insights or ang guidance.

Best regards,
Wei

**Reply by GiudGiud** [reply]
I see a lot of time being spent in *, times and tensor_product, which are operations on tensors
Any way to simplify the tensor math to have less operations? Are you using the most compact (symmetric?) representation there is?

**Reply by PengWei97** [reply]
Hi @GiudGiud,

I'm very sorry for taking so long to reply to you. In fact, I have found the reason why "Calculated Residual" is displayed for a long time in the first place. Since I set `maximum_substep_iteration = 100`，the loop of Line 316 in ComputeMultiCPStressFullyCoupled can reach an astonishing 2^100 times.

When I set `maximum_substep_iteration = 10`, the following message is displayed indicating that the calculation of the CP constitutive model failed, and there are two activated order parameters in the element where the constitutive calculation failed.

```bash
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
the id of current element: 7
the num of activate op: 2
the grain_id: 0
the num of substep: 1
the istep: 0
the num of substep: 2
the istep: 0
the num of substep: 4
the istep: 0
the num of substep: 8
the istep: 0
the num of substep: 16
the istep: 0
the num of substep: 32
the istep: 0
the num of substep: 64
the istep: 0
the num of substep: 128
the istep: 0
the num of substep: 256
the istep: 0
the num of substep: 512
the istep: 0

ComputeMultiCPStressFullyCoupled: Constitutive failure
To recover, the solution will fail and then be re-attempted with a reduced time step

A MooseException was raised during FEProblemBase::computeResidualTags
ComputeMultiCPStressFullyCoupled: Constitutive failure
To recover, the solution will fail and then be re-attempted with a reduced time step.

Nonlinear solve did not converge due to DIVERGED_FUNCTION_DOMAIN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge


*** ERROR ***
/home/pw-moose/projects/panda/simulation/prm3/step01_elastic_energy/s2_bicrystal_cp_elastic3.i:403.3:
The following error occurred in the TimeStepper 'IterationAdaptiveDT' of type IterationAdaptiveDT.

Solve failed and timestep already at dtmin, cannot continue!

Time Step 20, time = 0.15474, dt = 1e-12

Grain Tracker Status:
Grains active index 0: 1 -> 1
Grains active index 1: 1 -> 1

Finished inside of GrainTracker

ComputeMultiCPStressFullyCoupled: Constitutive failure
To recover, the solution will fail and then be re-attempted with a reduced time step.


A MooseException was raised during FEProblemBase::computeResidualTags
ComputeMultiCPStressFullyCoupled: Constitutive failure
To recover, the solution will fail and then be re-attempted with a reduced time step.

Nonlinear solve did not converge due to DIVERGED_FUNCTION_DOMAIN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge


*** ERROR ***
/home/pw-moose/projects/panda/simulation/prm3/step01_elastic_energy/s2_bicrystal_cp_elastic3.i:403.3:
The following error occurred in the TimeStepper 'IterationAdaptiveDT' of type IterationAdaptiveDT.

Solve failed and timestep already at dtmin, cannot continue!

Stack frames: 14
```

So, what is certain now is that the problem occurs during the constitutive calculation process. As for the specific reason, I am still looking for it.

I would greatly appreciate any insights or suggestions from the community to help resolve this issue.

Wei

**Reply by GiudGiud** [reply]
@dschwen @laagesen on advice on constitutive models

**Reply by laagesen** [reply]
Sorry, not in my wheelhouse. @sapitts or @jiangwen84 might be able to help

**Reply by sapitts** [reply]
Likely you have a state variable in the crystal plasticity model that is not converging properly. You will probably have to add in some numerical bounds/ guards on the variable(s) that are causing the problem. To start identifying the problematic variables, I'd recommend adding 
```
print_state_variable_convergence_error_messages = true
```
to both your stress input file block and your crystal plasticity constitutive model input file block. Please note this will give you *A LOT* of console output.

This new output will help you identify where in the crystal plasticity stress and/or constitutive model the simulation is struggling; you will likely still need to add variable specific print statements or use a debug to track the evolution of the variables updated in that part of the code

**Reply by PengWei97** [reply]
Hi @sapitts,

Your suggestion is greatly appreciated,. We've attempted to incorporate `print_state_variable_convergence_error_messages = true` within `CrystalPlasticityKalidindiUpdate`, and tried printing out key variables of elements where constitutive computations failed using print statements. The output to the terminal is as follows:

```
Time Step 28, time = 0.151255, dt = 0.001

Grain Tracker Status:
Grains active index 0: 1 -> 1
Grains active index 1: 1 -> 1

Finished inside of GrainTracker

ComputeMultiCPStressFullyCoupled::preSolveQp()
[_qp]: 0
_activate_op_index: 1
(*_vals[op_index])[_qp]: 0.00784244
(*_pk2_gr[_active_op_index])[_qp]: (xx,xy,xz)=(       0,        0,        0)
(yx,yy,yz)=(       0,        0,        0)
(zx,zy,zz)=(       0,        0,        0)

CPStressFullyCoupledUpdateBase::calculateShearStress
[_qp]: 0
pk2: (xx,xy,xz)=(       0,        0,        0)
(yx,yy,yz)=(       0,        0,        0)
(zx,zy,zz)=(       0,        0,        0)

_flow_direction: (xx,xy,xz)=(       0,        0,       -0)
(yx,yy,yz)=(0.408248, 0.408248, -0.408248)
(zx,zy,zz)=(0.408248, 0.408248, -0.408248)

_tau: 0
_slip_increment[_qp][i]: 0
_substep_dt: 0.001
_tau[_qp][i]: 0
(*_slip_resistance_gr[_active_op_index])[_qp][i]: 10

CPStressFullyCoupledUpdateBase::calculateShearStress
[_qp]: 0
pk2: (xx,xy,xz)=( 193.258, 1.67442e-06,        0)
(yx,yy,yz)=(1.67442e-06,  260.941,        0)
(zx,zy,zz)=(       0,        0,  190.268)

_flow_direction: (xx,xy,xz)=(       0,        0,       -0)
(yx,yy,yz)=(0.408248, 0.408248, -0.408248)
(zx,zy,zz)=(0.408248, 0.408248, -0.408248)

_tau: 28.8521
_slip_increment[_qp][i]: 1.59794e+06
_substep_dt: 0.001
_tau[_qp][i]: 28.8521
(*_slip_resistance_gr[_active_op_index])[_qp][i]: 10

*** Warning ***
/home/pengwei/projects/panda/simulation/prm3/step01_elastic_energy/s2_bicrystal_cp_elastic3.i:313.3:
The following warning occurred in the MaterialBase 'trial_xtalpl' of type CPKalidindiFullyCoupledUpdate.

Maximum allowable slip increment exceeded 1.59794e+06

ComputeMultiCPStressFullyCoupled: Constitutive failure
To recover, the solution will fail and then be re-attempted with a reduced time step.


A MooseException was raised during FEProblemBase::computeResidualTags
ComputeMultiCPStressFullyCoupled: Constitutive failure
To recover, the solution will fail and then be re-attempted with a reduced time step.

Nonlinear solve did not converge due to DIVERGED_FUNCTION_DOMAIN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge
```

The element is activated at Time Step 27 with only one active order parameter, hence only one constitutive model is active. However, at Time Step 28, another order parameter (gr1) is activated, requiring the calculation of a new crystal plasticity constitutive model with the same total deformation gradient. However, the output indicates that the slip increment in the newly created constitutive model far exceeds its maximum allowable value (0.02). We believe this discrepancy leads to the failure of constitutive solving.

This situation is akin to suddenly applying a large load or strain, rather than a gradually increasing strain rate load. To validate this hypothesis, we created a simplified case based on exception.i considering only crystal plasticity mechanics, where boundary conditions are set as follows:

```bash
  [tdisp]
    type = FunctionDirichletBC
    variable = uy
    boundary = top
    function = 'if(t<0.2, 0.0, 0.05)'
  []
```

This case still encounters the same issue and displays identical error messages. Thus, we believe this simple case confirms our suspicion that the newly created crystal plasticity model fails to converge numerically when subjected to a sudden large total deformation gradient, instead of a gradual strain rate. 

In conclusion, within crystal plasticity-phase field coupling simulations, when activating new order parameters, new crystal plasticity constitutive models need to be continuously created. While the total deformation gradient remains the same at each quadrature point, elastic and plastic deformation gradients, as well as internal state variables, vary. This was confirmed in discussion with Professor Michael R. Tonks. 

Given your extensive experience in writing crystal plasticity constitutive models, do you have any suggestions regarding this issue?

I would greatly appreciate any insights or suggestions from the community to help resolve this issue.

Wei

**Reply by sapitts** [reply]
Hi Wei,

First, please accept my apologies for my long delay in responding to your message.

I'm not surprised that the crystal plasticity code failed in the sudden-large-load case. The Kalidindi model is a fairly simple crystal plasticity model, and it likely does not have all the dislocation motion mechanisms necessary to accommodate a large sudden load. The crystal plasticity models that we have in main MOOSE are much more suited to lower loading rates because they are built using the quasi-static assumptions.

If you are still continuing with the simulation, a couple of options you might try, either separately or in combination:
 1. gradually apply the load to the crystal plasticity model
 2. change the tolerance on the maximum slip tolerance rate (I wouldn't be comfortable with a slip increment of 1e+6 personally, but you might increase the tolerance amount some to allow the simulation to run)

I know you had previously tried to apply suggestion 1. with substepping and ran into a hanging simulation. You might try to more explicitly trying this substepping by providing the crystal plasticity model simulation, maybe with a MultiApp approach that might allow the crystal plasticity model to take a *much* smaller timestep than the phase field model.

Hope this helps, or, that since I took so long to respond, that you have already found a suitable solution for this problem,
Stephanie

p.s. your Schmid tensor (the flow direction tensor) looks odd to me--have you checked to ensure that Schmid tensor calculation is giving the result you expect?
